### ISSUE TITLE
Mapsforge - Zoom level inconsistent between online and offline

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#7750]
We have some similar issues listed here somewhere, however while using todays nightly after upgrade to mf0.11 (2019.07.28-NB) I recognized a sudden big difference between online tiles and offline tiles. 
I am pretty sure the following described difference is either new or has become worse with mf0.11:

While the offline map rendering is pretty much as it should be speaking about zoom level, detail level, etc. the online tiles from OSM have a problem with maximum zoom size and/or font size and/or tile resolution.

My assumption would be, that somehow the online tiles are shown with a wrong DPI or there is some kind of mismatch between tile area and zoom control.

That is an area of Cologne shown using mf0.11 and openandromaps offline file for Germany:
![Screenshot_20190728-113130_cgeo](https://user-images.githubusercontent.com/949669/62012078-caa1a700-b181-11e9-97be-ce18d50da8e2.jpg)

This is the same area of Cologne shown using mf0.11 and OSM-online tiles (just changes the map source and zoomed in as much as possible):
![Screenshot_20190728-113059_cgeo](https://user-images.githubusercontent.com/949669/62012089-ec9b2980-b181-11e9-8d11-c57d88d5615a.jpg)
I would have expected to see the same detail level and zoomed area. If the server does not provide more detailed tiles, then it should be enough to just enlarge the existing tiles (they are shown quite small, i.e. I cannot read the text).

Google MapsV1 in c:geo with highest zoom as comparison:
![Screenshot_20190728-113115_cgeo](https://user-images.githubusercontent.com/949669/62012101-2704c680-b182-11e9-8a38-4869b0e9f350.jpg)

OSM online tiles in Locus as comparison:
![Screenshot_20190728-122408_Locus Map](https://user-images.githubusercontent.com/949669/62012114-3edc4a80-b182-11e9-867e-0da2c36e1830.jpg)

**Comment by Lineflyer** [comment]
I closed some older issues in preference for this one.

Maybe important information taken from #6149 from @rsudev:

After researching for a while, I think I now know the background for this issue.
The zoom level relates to the tile the map is drawn with.
- Google map always uses tile size 256
- Mapsforge online always uses tile size 256
- Mapsforge offline uses a tile size depending on the the resolution (I tested mdpi through to xxhdpi):
  - Minimum (up to hdpi) is 256
  - xhdpi 512
  - xxhdpi 768

In addition is there an offset of 1 between Google and Mapsforge (Google starts at 1, Mapsforge at 0).
The problem now is, that the zoom levels do not fully align, depending on the tile size. For Tile size 768, the zoom level difference is 1.5 (which is not possible to achieve as it is an integer).
I think I'll stick with rounding up in this case, let's see how it looks like in action than.

**Comment by jonas-koeritz** [comment]
This really looks like a density issue. I also noticed that when switching
between map sources.
I will have a look at this soon. The mapsforge 0.11 migration seems to have
become my main project.

Lars <notifications@github.com> schrieb am So., 28. Juli 2019, 21:54:

**Comment by jonas-koeritz** [comment]
There is also a problem with the zoom controls. When the Zoom+ button is already disable I can still zoom in with double tapping and pinch gestures on the online map. The map disappears when "overzoomed".

**Comment by Lineflyer** [comment]
> he mapsforge 0.11 migration seems to have
> become my main project.

Highly appreciated and thanks :)

**Comment by Lineflyer** [comment]
AFAIR we have an issue for this overzooming/empty map problem somewhere around.

**Comment by jonas-koeritz** [comment]
- Locus seems to somehow "overzoom" the OSM Tiles, the scale we see on our devices is exactly what OSM delivers
- Mapsforge seems to not request Zoom-Level 19 tiles (I will open an issue for this with mapsforge to clarify)
- The problem is not the inconsistency between zoom levels, it's the "clipping" that occurs, when using offline maps we can get way closer without hitting the max zoom level of the layer, switching to OSM clips the zoom level to 18 and zooms us out to that.

The problems that need to be addressed are:

1. Get higher zoom-level tiles from OSM (Mapsforge issue)
2. Allow overzooming of tiles or find another way to deal with HighDPI smartphones (For reference: https://github.com/mapsforge/mapsforge/blob/42da27cc110357ade1f8163dbcbdf2ae6f3f36ba/mapsforge-map/src/main/java/org/mapsforge/map/layer/download/TileDownloadLayer.java#L53)

**Comment by jonas-koeritz** [comment]
Deleted my last comment, working on the solution right now ;)

**Comment by Lineflyer** [comment]
@jonas-koeritz 
I gave the change a spin with builds based on `release` (which is mf0.9) and `master` (which is mf0.11) and it really solves this long time problem. Very nice finding!

I tested changing between the map sources incl. offline files and Google maps and found no crashes (as remembered by @rsudev to be the reason for this change). 
There are slight difference between the zoom levels of the different sources, but nothing which would interfere with the user being able to keep orientated on the map after changing the source.

So for me this seems perfectly solved.
